File: solutions/ruby/bird-count/2/bird_count.rb
# frozen_string_literal: true

# BirdCount class
class BirdCount
  def self.last_week
    [0, 2, 5, 3, 7, 8, 4]
  end

  def initialize(birds_per_day)
    @birds_per_day = birds_per_day
  end

  def yesterday
    @birds_per_day[-2]
  end

  def total
    @birds_per_day.sum
  end

  def busy_days
    @birds_per_day.count { |c| c >= 5 }
  end

  def day_without_birds?
    @birds_per_day.include?(0)
  end
end

File: solutions/elixir/bird-count/2/lib/bird_count.ex
defmodule BirdCount do
  def today([]), do: nil
  def today([day | _]), do: day

  def increment_day_count([]), do: [1]
  def increment_day_count([day | rest]), do: [day + 1 | rest]

  def has_day_without_birds?([]), do: false

  def has_day_without_birds?([day | rest]),
    do: if(day == 0, do: true, else: has_day_without_birds?(rest))

  def total([]), do: 0
  def total([first | rest]), do: first + total(rest)

  def busy_days([]),
    do: 0

  def busy_days([first | rest]),
    do: if(first >= 5, do: 1, else: 0) + busy_days(rest)
end
